ConfigureUtil deprecation warning

fyi I noticed this in a build scan,

> The org.gradle.util.ConfigureUtil type has been deprecated. 
This is scheduled to be removed in Gradle 9.0.
[Documentation](https://docs.gradle.org/7.6-rc-1/userguide/upgrading_version_7.html#org_gradle_util_reports_deprecations)

https://github.com/gradle-nexus/publish-plugin/blob/0aea95597e2eeaf9275a286254894bad29b3d1e1/src/main/kotlin/io/github/gradlenexus/publishplugin/DefaultNexusRepositoryContainer.kt#L32

> My latest invention is backwards and forwards compatible on all supported versions. It doesn't matter what Gradle the build is using, only the jar it produces.

Oh, I missed that fact that it requires 7.1, but should work also with 6.x.

I locally cherry-picked that change to master (and Gradle migration to 7.1) and from the regression tests, it works with Gradle 6.8+ [1], which is enough. However, I'm more a Groovy guy than a Kotlin one. Do you think, Gradle 7 and its migration to Kotlin 1.4 could affect 6.x users? Or we should rather use a reflection and still build that next release with Gradle 6.9?

---

[1] - for lower Gradle versions I had:
> Execution failed for task ':closeSonatypeStagingRepository'.
> 'void kotlin.jvm.internal.PropertyReference1Impl.<init>(java.lang.Class, java.lang.String, java.lang.String, int)'

or:
> Execution failed for task ':closeSonatypeStagingRepository'.
> Could not initialize class io.github.gradlenexus.publishplugin.AbstractTransitionNexusStagingRepositoryTask$transitionStagingRepo$retrier$1$1
> ...
> Caused by: java.lang.NoClassDefFoundError: Could not initialize class io.github.gradlenexus.publishplugin.AbstractTransitionNexusStagingRepositoryTask$transitionStagingRepo$retrier$1$1

--- a/src/compatTest/kotlin/io/github/gradlenexus/publishplugin/NexusPublishPluginTests.kt
+++ b/src/compatTest/kotlin/io/github/gradlenexus/publishplugin/NexusPublishPluginTests.kt
@@ -954,10 +954,6 @@ class NexusPublishPluginTests {
 
     private fun gradleRunner(vararg arguments: String): GradleRunner {
         val warnings = when {
-            // There's a deprecation in Gradle 7.6 that doesn't allow for passing builds.
-            // See https://github.com/gradle-nexus/publish-plugin/issues/152 for more details.
-            // This conditional branch of when should be removed once that issue is fixed.
-            gradleVersion >= GradleVersion.version("7.6") -> "all"
             // Failing only became an option at Gradle 5.6.
             gradleVersion >= GradleVersion.version("5.6") -> "fail"
             else -> "all"

--- a/src/main/kotlin/io/github/gradlenexus/publishplugin/DefaultNexusRepositoryContainer.kt
+++ b/src/main/kotlin/io/github/gradlenexus/publishplugin/DefaultNexusRepositoryContainer.kt
@@ -20,6 +20,7 @@ import groovy.lang.Closure
 import org.gradle.api.Action
 import org.gradle.api.NamedDomainObjectContainer
 import org.gradle.api.internal.NamedDomainObjectContainerConfigureDelegate
+import org.gradle.util.GradleVersion
 import java.net.URI
 import javax.inject.Inject
 
@@ -39,10 +40,28 @@ internal open class DefaultNexusRepositoryContainer @Inject constructor(
     }
 
     override fun configure(configureClosure: Closure<*>): NamedDomainObjectContainer<NexusRepository> =
-        @Suppress("DEPRECATION") // TODO https://github.com/gradle-nexus/publish-plugin/issues/152
-        org.gradle.util.ConfigureUtil.configureSelf(
-            configureClosure,
-            this,
-            NamedDomainObjectContainerConfigureDelegate(configureClosure, this)
-        )
+        if (GradleVersion.current().baseVersion < GradleVersion.version("7.6")) {
+            // Keep using the old API on old Gradle versions.
+            // It was deprecated in Gradle 7.1, but only from Gradle 7.6 it emits a deprecation warning.
+            // https://docs.gradle.org/current/userguide/upgrading_version_7.html#org_gradle_util_reports_deprecations
+            // Note: this will fail to compile when this project starts building on Gradle 9.0,
+            // at which point, this will need to be fully resolved,
+            // OR this call for older support will need to be removed OR reflective.
+            @Suppress("DEPRECATION")
+            org.gradle.util.ConfigureUtil.configureSelf(
+                configureClosure,
+                this,
+                NamedDomainObjectContainerConfigureDelegate(configureClosure, this)
+            )
+        } else {
+            // Keep using the new *internal* API on new Gradle versions.
+            // At least until https://github.com/gradle/gradle/issues/23874 is resolved.
+            // Introduced in Gradle 7.1, it's internal but stable up until the latest Gradle 8.0 at the time of writing.
+            // The Gradle 7.1 version of this class is a verbatim copy of Gradle 8.0's version, but without the nagging.
+            org.gradle.util.internal.ConfigureUtil.configureSelf(
+                configureClosure,
+                this,
+                NamedDomainObjectContainerConfigureDelegate(configureClosure, this)
+            )
+        }
 }

